### PR TITLE
KTOR-5359 Add engine shutdown configuration

### PR DESCRIPTION
**Pull request by marychatte**
**Subsystem**
Server

**Motivation**
[KTOR-5359](https://youtrack.jetbrains.com/issue/KTOR-5359)

### CODE REVIEW

File: ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt [hunk 1]
```diff
@@ -37,6 +37,16 @@ public interface ApplicationEngine {
          * Specifies the minimum size of a thread pool used to process application calls.
          */
         public var callGroupSize: Int = parallelism
+
+        /**
+         * Specifies the maximum amount of time for activity to cool down in milliseconds.
+         */
+        public var gracePeriodMillis: Long = 500
```
Comment (rsinukov) [inline ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt:44]: I don't see any usages of it. Why make it public property?

File: ktor-server/ktor-server-cio/jvmAndNix/src/io/ktor/server/cio/EngineMain.kt [hunk 2]
```diff
@@ -19,8 +19,12 @@ public object EngineMain {
     public fun main(args: Array<String>) {
         val applicationEnvironment = commandLineEnvironment(args)
         val engine = CIOApplicationEngine(applicationEnvironment) { loadConfiguration(applicationEnvironment.config) }
+        val gracePeriod =
+            engine.environment.config.propertyOrNull("ktor.deployment.gracePeriod")?.getString()?.toLong() ?: 50
+        val timeout =
+            engine.environment.config.propertyOrNull("ktor.deployment.timeout")?.getString()?.toLong() ?: 5000
```
Comment (rsinukov) [inline ktor-server/ktor-server-cio/jvmAndNix/src/io/ktor/server/cio/EngineMain.kt:25]: `ktor.deployment.timeout` is to generic name. can you specify that it's the timeout for stopping a server? same for `gracePeriod`

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt (hunk 1 above)
Comment (marychatte) [reply ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt:44]: It is used in `public fun BaseApplicationEngine.Configuration.loadCommonConfiguration`

Comment (rsinukov) [reply ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt]: But it's only set there and not read anywhere

Comment (marychatte) [reply ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt]: But why is it bad that it is public?

Comment (rsinukov) [reply ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt]: I don't understand why it should be. There is nothing users can do with it, and we don't use it as well, only set a value to it.

Comment (marychatte) [reply ktor-server/ktor-server-host-common/jvmAndNix/src/io/ktor/server/engine/ApplicationEngine.kt]: Understand, I agree that it is unnecessary

**Review by rsinukov — APPROVED** [review]

**Pull request merged**